src/IO/ReadBufferFromS3.cpp

```diff
@@ -74,7 +91,10 @@ bool ReadBufferFromS3::nextImpl()
     if (read_until_position)
     {
         if (read_until_position == offset)
+        {
+            read_all_range_successfully = true;
```

Comment:
But AsynchronousBoundedReadBuffer has the same check, which means that for the last read range we will never get to this point (as well as to line 178) in code even though the range was all read successfully. I think we can add the same check in destructor before resetting the session and update read_all_range_successfully?

Reply:
yep, it is a good idea